sound/soc/sof/topology.c

```diff
@@ -2394,6 +2387,17 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 		return ret;
 	}
 
+	/*
+	 * DAI links are expected to have exactly 1 hw_config.
+	 * But some older topologies might have no hw_config for HDA dai links.
+	 */
+	if (le32_to_cpu(cfg->num_hw_configs) != 1 &&
```

Comment:
that's not correct. dai links could have more than one config, and in fact it's an issue that we've pushed under the rug consistently.
The code should be 
numconfig == 0 && config.type != SOF_DAI_INTEL_HDA.

what happens btw if we have a topology with no HDAudio config, would things work or would we generate an oops?

Reply:
@plbossart if we can have more than 1 hw_config, we dont have the logic to process them here. How would we handle that?

To answer you second question, if there are no HDA configs in topology,i cant imagne what would cause a kernel oops.

Reply:
well multiple configs would be another "topology ABI" bump triggering a different behavior on newer kernels.
For the second part it'd like to see a test.

Reply:
> well multiple configs would be another "topology ABI" bump triggering a different behavior on newer kernels.

@plbossart OK, fixed now. 
> For the second part it'd like to see a test.

I tested the sof-pcm-521x topology without the HDA pipelines and config. I didnt see any oops.

Reply:
ok, that's fine then